### PR TITLE
Time - Hannah J

### PR DESCRIPTION
**Pull request by Hannah0124**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | `Model` contains data and state of the application. It is independent of UI, so it is reusuable.
Describe in your own words what the Controller is doing in Rails | `Controller` is like a main processor between `model` and `view`. It receives a request (events) from a browser and retreives data from a `model` and passes the data to the `view`.
Describe in your own words what the View is doing in Rails | `View` (erb files) changes data to HTML. In other words, `View` turns into UI. It only contains display logic.
Describe an edge-case controller test you wrote | For the `update` action, if an invalid id (e.g. -1) was given, it tests (1) if `Task.count` does not change and (2) if it's redirectd to the root page.
What is the purpose of using strong params? (i.e. the params method in the controller) | `Strong params` provides a safe way to require and permit data that comes from our forms.
How are Rails migrations related to Rails models? | `Migration` provides a blueprint of (1) how to construct the database schema and (2) history of the database. Migrations add or remove tables, columns and rows in `Model`.
Describe one area of Rails that are still unclear on | I think the big picture of Rails is still not clear to me. I beleive that over projects, I will have a better understanding.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# Reference: https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/08-rails/routes-and-resources.md
+Rails.application.routes.draw do
+  resources :tasks
+  root to: "tasks#index"  # root_path
+  patch '/tasks/:id/complete', to: 'tasks#complete', as: 'complete_task'
```
Comment (beccaelenzil) [inline config/routes.rb:5]: Consider how making two separate methods, for instance `mark_complete` and `mark_incomplete` would make this action idempotent.

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -58,10 +62,9 @@
     end
   end
   
-  describe "create" do
+
+  describe "create action" do
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:66]: These tests look great. Nice work implementing this optional feature. Remember to include an expectation for where it should redirect to.

File: test/controllers/tasks_controller_test.rb [hunk 3]
```diff
@@ -85,40 +88,211 @@
...
     
     it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
       # Your code here
+      # Act 
+      get edit_task_path(-1)
+
+      # Assert 
+      must_respond_with :redirect
+      must_redirect_to root_path
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:111]: You do not need to also include the assertion on line 110. `must_redirect_to root_path` is enough :)

File: app/assets/stylesheets/tasks.scss [hunk 4]
```diff
@@ -0,0 +1,249 @@
+// Place all the styles related to the Tasks controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
+
+@import url('https://fonts.googleapis.com/css2?family=Dancing+Script:wght@700&display=swap');
+
+@import url('https://fonts.googleapis.com/css2?family=Caveat:wght@700&display=swap');
+
+body {
```
Comment (beccaelenzil) [inline app/assets/stylesheets/tasks.scss:16]: Great work creating dynamic styling.

File: app/views/tasks/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+<%# form reference: https://m.patrikonrails.com/rails-5-1s-form-with-vs-old-form-helpers-3a5f72a8c78a%>
+
+<%# https://api.rubyonrails.org/classes/ActionView/Helpers/FormHelper.html#method-i-check_box%>
+<% if @tasks.length > 0 %>
+  <ul class='tasks-container'>
+    <% @tasks.each do |task| %>
+      <li class='one-task-container'>
+        
+        <%= form_with model: @task, class: 'task-complete', url: complete_task_path(task.id), method: :patch do |f| %>
+          <%= f.check_box :completed_at, { class: "check-box", onChange: 'form.submit()', checked: !task.completed_at.nil? } %>
```
Comment (beccaelenzil) [inline app/views/tasks/index.html.erb:14]: Great work implementing the checkbox feature.

**Comment by beccaelenzil** [comment]
# Task List




## Major Learning Goals/Code Review





| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| At least 6 commits with meaningful commit messages | ✔️
| Routes follow RESTful conventions | ✔️
| Uses named routes (like `_path`) | ✔️
| Creates Models and migrations | ✔️
| Creates styled views | ✔️ WOW! Nice stylin'
| Handles errors like nonexistant tasks | ✔️
| Uses `form_with` to render forms in Rails | ✔️

## Functional Requirements/Manual Testing

| Functional Requirement | yes/no |
| --- | --- |
| Successfully handles index & show | ✔️
| index & show tests pass | ✔️
| Successfully handles: New, Create | ✔️
| New, Create tests pass | ✔️
| Successfully handles: Edit, Update | ✔️
| Successfully handles: Destroy, Task Complete | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 5+ in Functional Requirements | ✔️
 





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

**Review by beccaelenzil — COMMENTED** [review]
Great work on this assignment! It is clear that the learning goals were me and you're off to a great start on rails. In particular, nice work on the styles and implementing tests for the create, update, destroy, and complete actions. Keep up the hard work!
